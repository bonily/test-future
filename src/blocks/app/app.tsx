/* eslint-disable no-console */
import React from 'react';
import Persons from '../persons/persons';
import {PersonType} from '../../types';
import {ActionCreator, AppStateType, Operation} from '../../reducer';
import {connect} from 'react-redux';
import StartWindow from '../start-window/start-window';
import {getSortedPerson, getSortType} from '../../selector';
import Header from '../header/header';
import {Values} from '../person-form/person-form';


interface MapStatePropsType {
  persons: PersonType[],
  sortType: string,
  error: string | number,
  inputValue: string,
  isStart: boolean
}


interface MapDispatchToPropsType {
  onSortTitleClick: (sortType: string, isAscending: boolean) => void,
  onInputChange: (arg0: string) => void,
  onDataSizeButtonClick: (sizeType: string) => void,
  setError: (error: string) => void,
  onAddNewPersonButtonClick: (value : Values) => void
}


const App: React.FunctionComponent<MapStatePropsType & MapDispatchToPropsType> = (props: MapStatePropsType & MapDispatchToPropsType) => {
  const {error, persons, sortType, isStart, onDataSizeButtonClick, onSortTitleClick, onInputChange, onAddNewPersonButtonClick, setError} = props;

  return (
    isStart && persons.length === 0 ?
      <StartWindow
        error = {error}
        onDataSizeButtonClick = {onDataSizeButtonClick}
        setError = {setError}
      /> :
      <><Header
        onInputChange={onInputChange}
        onAddNewPersonButtonClick = {onAddNewPersonButtonClick}/>
      <Persons
        persons={persons}
        sortType={sortType}
        onSortTitleClick={onSortTitleClick}
      /></>
  );
};

const mapStateToProps = (state: AppStateType): MapStatePropsType => ({
  persons: getSortedPerson(state),
  sortType: getSortType(state),
  error: state.error,
  inputValue: state.inputValue,
  isStart: state.isStart
});


const mapDispatchToProps = (dispath : any): MapDispatchToPropsType => ({
  onDataSizeButtonClick(sizeType) {
    dispath(Operation.loadPersons(sizeType));
  },
  onSortTitleClick(sortType, isAscending) {
    dispath(Operation.changeSortType(sortType, isAscending));
  },
  onInputChange(str) {
    dispath(ActionCreator.filterPersons(str));
  },
  setError(error) {
    dispath(ActionCreator.setError(error));
  },
  onAddNewPersonButtonClick(value) {
    dispath(ActionCreator.addNewPerson(value));
  }
});


export {App};
export default connect(mapStateToProps, mapDispatchToProps)(React.memo(App));
