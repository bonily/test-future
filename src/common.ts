import {PersonType} from './types';
import {Values} from './blocks/person-form/person-form';
import {SORT_TYPES} from "./const";

export const extend = (a : any, b: any): any => {
  return Object.assign({}, a, b);
};

export const getPersonsBySortType = (persons : PersonType[], sortType : string, ascending : boolean): PersonType[] => {
  switch (sortType) {
    case SORT_TYPES.FIRST_NAME:
      return persons.sort((a, b) => ascending ? a.firstName.localeCompare(b.firstName) : b.firstName.localeCompare(a.firstName)).slice();
    case SORT_TYPES.LAST_NAME:
      return persons.sort((a, b) => ascending ? a.lastName.localeCompare(b.lastName) : b.lastName.localeCompare(a.lastName)).slice();
    case SORT_TYPES.EMAIL:
      return persons.sort((a, b) => ascending ? a.email.localeCompare(b.email) : b.email.localeCompare(a.email)).slice();
    case SORT_TYPES.ID:
      return persons.sort((a, b) => ascending ? a.id - b.id : b.id - a.id).slice();
    case SORT_TYPES.PHONE:
      return persons.sort((a, b) => ascending ? a.phone.localeCompare(b.phone) : b.phone.localeCompare(a.phone)).slice();
  }
  return persons.slice();
};

export const noop = () : void => {
  // do nothing
};

export const getFilteredPersons = (persons : PersonType[], str = ``):PersonType[] => {
  return persons.filter((person) =>
    person.firstName.toLowerCase().indexOf(str) > -1 ||
    String(person.id).indexOf(str) > -1 ||
    person.lastName.toLowerCase().indexOf(str) > -1 ||
    person.email.toLowerCase().indexOf(str) > -1 ||
    person.phone.toLowerCase().indexOf(str) > -1 ||
    person.address.streetAddress.toLowerCase().indexOf(str) > -1 ||
    person.address.state.toLowerCase().indexOf(str) > -1 ||
    person.address.zip.toLowerCase().indexOf(str) > -1 ||
    person.address.city.toLowerCase().indexOf(str) > -1);
};

// export const updateFavoriteStatus = (persons : PersonType[], id : number):PersonType[] => {
//   const index = persons.findIndex((person) => person.id === id);
//   persons[index].favourite = !persons[index].favourite;
//   return persons;
// };

export const castFormValueToPersonType = (value: Values) : PersonType => {
  return {
    id: Number(value.id),
    firstName: value.firstName,
    lastName: value.lastName,
    email: value.email,
    phone: value.phone,
    address: {
      streetAddress: value.streetAddress,
      city: value.city,
      state: value.state,
      zip: value.zip
    },
    description: value.description,
  };
};
