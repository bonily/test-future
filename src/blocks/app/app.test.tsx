import React from "react";
import * as renderer from "react-test-renderer";
import {noop} from "../../common";
import {App} from "./app";


const persons = [
  {
    address: {
      city: `Bradford`,
      state: `TX`,
      streetAddress: `7426 Suspendisse Dr`,
      zip: `14783`,
    },
    description: `pulvinar magna convallis quis porta adipiscing mi porta vestibulum elit amet ipsum egestas tortor suspendisse mattis rutrum lacus vel libero vitae sit dolor molestie lorem sed massa lacus lacus vitae placerat et`,
    email: `MRamsey@scelerisque.com`,
    firstName: `Jeri`,
    id: 714,
    lastName: `Wittcop`,
    phone: `(731)368-3302`
  },
  {
    address: {
      city: `Bradford`,
      state: `TX`,
      streetAddress: `7426 Suspendisse Dr`,
      zip: `14745`,
    },
    description: `pulvinar magna convallis quis porta adipiscing mi porta vestibulum elit amet ipsum egestas tortor suspendisse mattis rutrum lacus vel libero vitae sit dolor molestie lorem sed massa lacus lacus vitae placerat et`,
    email: `MRamsey@scelerisque.com`,
    firstName: `Bonya`,
    id: 120,
    lastName: `Wittcop`,
    phone: `(731)368-3302`
  }
];


it(`App on the start should render start window `, () => {
  const tree = renderer
    .create(
        <App
          persons={[]}
          inputValue={``}
          sortType={``}
          isStart={true}
          error={``}
          onAddNewPersonButtonClick={noop}
          onDataSizeButtonClick={noop}
          onSortTitleClick={noop}
          onInputChange={noop}
          setError={noop}
        />
    ).toJSON();

  expect(tree).toMatchSnapshot();
});

it(`App with persons should render table block `, () => {
  const tree = renderer
    .create(
        <App
          persons={persons}
          inputValue={``}
          sortType={``}
          isStart={false}
          error={``}
          onAddNewPersonButtonClick={noop}
          onDataSizeButtonClick={noop}
          onSortTitleClick={noop}
          onInputChange={noop}
          setError={noop}
        />
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
