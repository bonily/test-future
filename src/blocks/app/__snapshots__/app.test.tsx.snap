// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App on the start should render start window  1`] = `
<div
  className="sc-kEjbxe kuoOcw"
>
  <button
    className="sc-iqHYGH kwhEwP"
    onClick={[Function]}
  >
    Полный список
  </button>
  <button
    className="sc-iqHYGH kwhEwP"
    onClick={[Function]}
  >
    Короткий список
  </button>
</div>
`;

exports[`App with persons should render table block  1`] = `
Array [
  " ",
  <form
    className="sc-hBEYos imFMyH"
  >
    <input
      className="sc-fodVxV iJMQbK"
      onChange={[Function]}
      placeholder="Поиск"
      type="text"
    />
    <button
      className="sc-fFubgz JKHnB"
      name="find-button"
      onClick={[Function]}
      type="submit"
    >
      Найти
    </button>
    <button
      className="sc-fFubgz JKHnB"
      id="new-person"
      onClick={[Function]}
      type="button"
    >
      Добавить
    </button>
  </form>,
  <table
    className="sc-gKsewC dCXYBn"
  >
    <colgroup>
      <col
        style={
          Object {
            "background": "yellow",
          }
        }
      />
      <col />
      <col />
      <col />
      <col />
    </colgroup>
    <tbody>
      <tr
        className="sc-iBPRYJ"
      >
        <th
          className="sc-eCssSg hBBrPS"
          onClick={[Function]}
        >
          id
          <img
            className="sc-jSgupP diBDkZ"
            height="18px"
            src="./img/sort-icon.svg"
            width="18px"
          />
        </th>
        <th
          className="sc-eCssSg hBBrPS"
          onClick={[Function]}
        >
          firstName
          <img
            className="sc-jSgupP diBDkZ"
            height="18px"
            src="./img/sort-icon.svg"
            width="18px"
          />
        </th>
        <th
          className="sc-eCssSg hBBrPS"
          onClick={[Function]}
        >
          lastName
          <img
            className="sc-jSgupP diBDkZ"
            height="18px"
            src="./img/sort-icon.svg"
            width="18px"
          />
        </th>
        <th
          className="sc-eCssSg hBBrPS"
          onClick={[Function]}
        >
          email
          <img
            className="sc-jSgupP diBDkZ"
            height="18px"
            src="./img/sort-icon.svg"
            width="18px"
          />
        </th>
        <th
          className="sc-eCssSg hBBrPS"
          onClick={[Function]}
        >
          phone
          <img
            className="sc-jSgupP diBDkZ"
            height="18px"
            src="./img/sort-icon.svg"
            width="18px"
          />
        </th>
      </tr>
      <tr
        className="sc-iBPRYJ"
        onClick={[Function]}
      >
        <td
          className="sc-fubCfw gJkPIV"
        >
          714
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          Jeri
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          Wittcop
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          MRamsey@scelerisque.com
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          (731)368-3302
        </td>
      </tr>
      <tr
        className="sc-iBPRYJ"
        onClick={[Function]}
      >
        <td
          className="sc-fubCfw gJkPIV"
        >
          120
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          Bonya
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          Wittcop
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          MRamsey@scelerisque.com
        </td>
        <td
          className="sc-fubCfw gJkPIV"
        >
          (731)368-3302
        </td>
      </tr>
    </tbody>
  </table>,
  <nav
    className="sc-bdfBwQ cvWBOr"
  >
    <ul
      className="sc-gsTCUz kIEiKe"
    >
      <li
        className="sc-dlfnbm iKQuvh"
      >
        <a
          className="sc-hKgILt gWxVff"
          href="#!"
          onClick={[Function]}
        >
          1
        </a>
      </li>
    </ul>
  </nav>,
]
`;
